content/en/blog/_posts/2023-12-05-image-filesystem.md

```diff
@@ -0,0 +1,170 @@
+
+**Logs**
+
+Kubernetes stores the logs of each container running in `/var/logs`.  These logs are ephemeral and are monitored by Kubelet to make sure that they do not grow too large while the pods are running.
+
+**Container runtime**
+
+The container runtime has two different areas of storage for containers and images. Images are usually denoted as the read-only layer, as they are not modified when containers are running. There is a thin layer on top of containers that provides ephemeral storage for containers if the container is writing to the filesystem. This is referred to as the writable layer.
+
+## Container Runtime Configurations
+
+### CRI-O
+
+CRI-O uses a storage configuration file to control how the container runtime stores persistent and temporary data.
+CRI-O utilizes the [storage library](https://github.com/containers/storage).  
+Some Linux distributions have a manual entry for storage (`man 5 containers-storage.conf`).
+The main configuration for storage is located in `/etc/containers/storage.conf` and one can control the location for temporary data and the root directory.  
+The root directory is where CRI-O stores the persistent data.
+
+```toml
+[storage]
+# Default storage driver
+driver = "overlay"
+# Temporary storage location
+runroot = "/var/run/containers/storage"
+# Primary read/write location of container storage 
+graphroot = "/var/lib/containers/storage"
+```
+
+- graphroot
+  - Persistent data stored from the container runtime
+  - If SELinux is enabled, this must match the `/var/lib/containers/storage`
+- runroot
+  - Temporary read/write access for container
+  - Recommended to have this on a temporary filesystem
+
+A quick way to relabel your graphroot directory to match `/var/lib/containers/storage`.
+
+```bash
+semanage fcontext -a -e /var/lib/containers/storage /YOUR_STORAGE_PATH
+restorecon -R -v /YOUR_STORAGE_PATH
+```
+
+### containerd
+
+containerd uses a toml configuration file to control where persistent and ephemeral data is stored.
+The default path for the config file is located at `/etc/containerd/config.toml`.
+
+The relevant fields for containerd storage are `root` and `state`.
+
+- root
+  - The root directory for containerd metadata
+  - Default is `/var/lib/containerd`
+  - Root also requires selinux labels if your OS requires it
+- state
+  - Temporary data for containerd
+  - Default is `/run/containerd`
+
+## Kubernetes Node Pressure Eviction
+
+Kubernetes will automatically detect if the container filesystem is split from the node filesystem. When one separates the filesystem, Kubernetes is responsible for monitoring both the node filesystem and the container runtime filesystem.
+Kubernetes documentation refers to the node filesystem and the container runtime filesystem as nodefs and imagefs.
```

Comment:
This is true; that _is_ what the docs say right now, but it misses out on relevant detail. What if `/var` is not the same filesystem as `/`? Or how about having a mount point for `/var/lib/kubelet`? Or `/var/lib/kubelet/pods/`?

If all of those mount points existed, could a reader work out which one represents the nodefs?
Alternatively, is having more than two block-backed filesystems (`/` and the imagefs) undefined behavior?

_Ideally_, this article covers enough of that to work. If you want to make a docs tweak in support of this, you can do that and then you can link to the revised page.

Reply:
>We came to the conclusion that Kubelet will consider which disk `/var/lib/kubelet` is on to be the root filesystem.

OK, so: in this article, you should clarify that what Kubernetes calls the “root filesystem” is different from what the rest of the POSIX-speaking world uses the term to mean (ie: the filesystem that `/` is on).

We should also clarify that in the docs or possibly ask SIG Node to rethink the terminology, but that's something for a separate piece of work.

Reply:
Added a sentence about this. PTAL.

Reply:
We can merge this, but I think it's just barely OK and will still confuse readers.
See https://github.com/kubernetes/website/pull/43971/files#r1451741131 for a further suggestion.

Reply:
Added more clarification.